framework/mesh/MeshGenerator/MeshGenerator.h

```diff
@@ -40,30 +40,63 @@ class MeshGenerator : public ChiObject
   static chi::InputParameters GetInputParameters();
   explicit MeshGenerator(const chi::InputParameters& params);
 
-protected:
   /**Virtual method to generate the unpartitioned mesh for the next step.*/
   virtual std::unique_ptr<UnpartitionedMesh>
   GenerateUnpartitionedMesh(std::unique_ptr<UnpartitionedMesh> input_umesh);
 
+  struct VertexListHelper
+  {
+    virtual const chi_mesh::Vertex& at(uint64_t vid) const = 0;
+  };
+  template <typename T>
+  struct STLVertexListHelper : public VertexListHelper
+  {
+    explicit STLVertexListHelper(const T& list) : list_(list) {}
+    const chi_mesh::Vertex& at(uint64_t vid) const override
+    {
+      return list_.at(vid);
+    };
+    const T& list_;
+  };
+
+protected:
   // 01
+  /**Builds a cell-graph and executes the partitioner that assigns cell
+   * partition ids based on the supplied number of parts.*/
+  std::vector<int64_t> PartitionMesh(const UnpartitionedMesh& input_umesh,
+                                     int num_parts);
+
   /**Executes the partitioner and configures the mesh as a real mesh.*/
-  virtual std::shared_ptr<MeshContinuum>
-  SetupMesh(std::unique_ptr<UnpartitionedMesh> input_umesh_ptr);
+  std::shared_ptr<MeshContinuum>
+  SetupMesh(std::unique_ptr<UnpartitionedMesh> input_umesh_ptr,
+            const std::vector<int64_t>& cell_pids);
 
   // 02 utils
+  /**Broadcasts PIDs to other locations.*/
+
+  static void BroadCastPIDs(std::vector<int64_t>& cell_pids,
+                            int root,
+                            MPI_Comm communicator);
   /**Determines if a cells needs to be included as a ghost or as a local cell.*/
   bool
-  CellHasLocalScope(const chi_mesh::UnpartitionedMesh::LightWeightCell& lwcell,
+  CellHasLocalScope(int location_id,
```

Comment:
There are local cells and ghost cells. Both have local scope. This routine determines if a cell is either a local cell or a ghost, OR, if the mesh is replicated, then this will always return true.